go-controller/pkg/util/dns.go

```diff
@@ -0,0 +1,265 @@
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	// defaultTTL is used if an invalid or zero TTL is provided.
+	defaultTTL = 30 * time.Minute
+)
+
+type dnsValue struct {
+	// All IPv4 addresses for a given domain name
+	ips []net.IP
+	// Time-to-live value from non-authoritative/cached name server for the domain
+	ttl time.Duration
+	// Holds (last dns lookup time + ttl), tells when to refresh IPs next time
+	nextQueryTime time.Time
+}
+
+type DNS struct {
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+
+	// DNS resolvers
+	nameservers []string
+	// DNS port
+	port string
+}
+
+func NewDNS(resolverConfigFile string) (*DNS, error) {
+	config, err := dns.ClientConfigFromFile(resolverConfigFile)
+	if err != nil || config == nil {
+		return nil, fmt.Errorf("cannot initialize the resolver: %v", err)
+	}
+
+	return &DNS{
+		dnsMap:      map[string]dnsValue{},
+		nameservers: filterIPv4Servers(config.Servers),
+		port:        config.Port,
+	}, nil
+}
+
+func (d *DNS) Size() int {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return len(d.dnsMap)
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) GetIPs(dns string) []net.IP {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data.ips
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.dnsMap[dns] = dnsValue{}
+	_, err := d.updateOne(dns)
+	if err != nil {
+		delete(d.dnsMap, dns)
+	}
+
+	return err
+}
+
+func (d *DNS) Delete(dns string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	delete(d.dnsMap, dns)
+}
+
+func (d *DNS) Update(dnsName string) (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return d.updateOne(dnsName)
+}
+
+func (d *DNS) updateOne(dns string) (bool, error) {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		return false, fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+	}
+
+	ips, ttl, err := d.getIPsAndMinTTL(dns)
+	if err != nil {
+		res.nextQueryTime = time.Now().Add(defaultTTL)
+		d.dnsMap[dns] = res
+		return false, err
+	}
+
+	changed := false
+	if !ipsEqual(res.ips, ips) {
+		changed = true
+	}
+	res.ips = ips
+	res.ttl = ttl
+	res.nextQueryTime = time.Now().Add(res.ttl)
+	d.dnsMap[dns] = res
+	return changed, nil
+}
+
+func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
```

Comment:
You need to retrieve `AAAA` records here as well.

Reply:
added getting AAAA records